src/Database/LSMTree/Internal/Merge.hs

```diff
@@ -75,72 +84,119 @@ data MergeState =
 
 -- | Merges can either exist on a level of the LSM, or be a union merge of two
 -- tables.
+class IsMergeType t where
+  -- | A last level merge behaves differently from a mid-level merge: last level
+  -- merges can actually remove delete operations, whereas mid-level merges must
+  -- preserve them.
+  isLastLevel :: t -> Bool
+  -- | Union merges follow the semantics of @Data.Map.unionWith (<>)@. Since the
+  -- input runs are semantically treated like @Data.Map@s, deletes are ignored
+  -- and inserts act like mupserts, so they need to be merged monoidally using
+  -- 'resolveValue'.
+  isUnion :: t -> Bool
```

Comment:
This is now like the prototype. Doesn't make a big difference, but are these nicer to have than `toMergeType :: t -> MergeType`? If not, I can drop the last commit and we should probably change the prototype, too.